---
title: "Setting up"
engine: knitr
format:
    html:
        fig-width: 5
        fig-height: 4
        code-annotations: hover
---

All bioinformatics projects requires setting up the required software. Let's go through the setup for this RNA-seq workshop here.

## From reads to alignments

![](../images/flow-chart-RNA-Seq-2.png)

The first part is the 'hardcore' bioinformatics part of this workshop. These steps are nearly always performed using **command-line programs**, which can be accessed via the Terminal (macOS, Linux) or WSL (Windows). For this workshop, we prepared a Google Colab environment where we will run the commands. Access to the Google Colab environment requires a Google account.
    
* [Empty Google Colab environment](https://colab.research.google.com/drive/1jR2Og_qrUVg-VXNsfqJc5eEILcqHJYAC?usp=share_link): to write the commands yourself
* [Pre-filled Google Colab environment](https://colab.research.google.com/drive/1jR2Og_qrUVg-VXNsfqJc5eEILcqHJYAC?usp=share_link): to execute prewritten commands

In a real experiment, you will often have so many samples and reads that you will need access to a compute cluster with more compute power (CPUs), working memory (RAM), and storage space. At the University of Amsterdam, we use [Crunchomics](https://crunchomics-documentation.readthedocs.io/en/latest/). Crunchomics comes with many popular bioinformatics pre-installed, additional programs can be installed using `conda`.

## Differential gene expression with DESeq2

![](../images/flow-chart-RNA-Seq-3.png)

For the second of the workshop, we will use `R`. Please make sure you have `R` and `RStudio` installed on your computer. In addition, you will need to install the following packages, via two different routes:

Install the following packages via `install.packages()`

```{R filename = "R"}
#| eval: false 
install.packages('tidyverse')
install.packages('patchwork')
install.packages("pheatmap")
install.packages("BiocManager")
```

Install the following packages via `Bioconductor`

```{R filename = "R"}
#| eval: false 
BiocManager::install("apeglm")
BiocManager::install("clusterProfiler")
BiocManager::install("org.At.tair.db")
BiocManager::install("DESeq2")
```

Test whether `DESeq2` was installed properly by loading it:

```{R filename = "R"}
#| eval: false 
library(DESeq2)
```

During the workshop, we will **not** have time to install the packages, or to troubleshoot installation problems. Please try to install the packages before the workshop. You can always contact the [workshop organizers](https://www.biodsc.nl/people.html) if you run into problems.

## Getting the data

The reads and reference that will be used in this workshop can be found in the [following Github repository](https://github.com/mishapaauw/RNA-seq-workshop-materials), and will be downloaded into the Google Colab automatically.

The count table and experimental design file used in the DESeq2 part of this workshop can be found [here](https://github.com/mishapaauw/RNA-seq-workshop/tree/main/data).